File: demo-app/src/main/java/cn/dozyx/template/view/ViewLifeActivity.java
package cn.dozyx.template.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import cn.dozyx.template.R;

public class ViewLifeActivity extends AppCompatActivity {
    private static final String TAG = "ViewLifeActivity";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_view_life);
    }

    public static class ViewLifeView extends View {
        public ViewLifeView(Context context) {
            super(context);
            Log.d(TAG, "ViewLifeView: constructor1");
        }

        public ViewLifeView(Context context,
                @Nullable AttributeSet attrs) {
            super(context, attrs);
            Log.d(TAG, "ViewLifeView: constructor2");
        }

        public ViewLifeView(Context context,
                @Nullable AttributeSet attrs, int defStyleAttr) {
            super(context, attrs, defStyleAttr);
            Log.d(TAG, "ViewLifeView: constructor3");
        }

        @Nullable
        @Override
        protected Parcelable onSaveInstanceState() {
            Log.d(TAG, "onSaveInstanceState: ");
            return super.onSaveInstanceState();
        }

        @Override
        protected void onFinishInflate() {
            super.onFinishInflate();
            Log.d(TAG, "onFinishInflate: ");
        }

        @Override
        protected void onAttachedToWindow() {
            super.onAttachedToWindow();
            Log.d(TAG, "onAttachedToWindow: ");
        }

        @Override
        protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
            super.onLayout(changed, left, top, right, bottom);
            Log.d(TAG, "onLayout: ");
        }

        @Override
        protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec);
            Log.d(TAG, "onMeasure: ");
        }

        @Override
        protected void onDraw(Canvas canvas) {
            super.onDraw(canvas);
            Log.d(TAG, "onDraw: ");
        }

        @Override
        protected void onDetachedFromWindow() {
            super.onDetachedFromWindow();
            Log.d(TAG, "onDetachedFromWindow: ");
        }

        @Override
        protected void onFocusChanged(boolean gainFocus, int direction,
                @Nullable Rect previouslyFocusedRect) {
            super.onFocusChanged(gainFocus, direction, previouslyFocusedRect);
            Log.d(TAG, "onFocusChanged: " + gainFocus);
        }

        @Override
        protected void onRestoreInstanceState(Parcelable state) {
            super.onRestoreInstanceState(state);
            Log.d(TAG, "onRestoreInstanceState: ");
        }
    }

    public static class LifeViewLinearLayout extends LinearLayout {
        public LifeViewLinearLayout(Context context) {
            super(context);
            Log.d(TAG, "LifeViewGroup: 1");
        }

        public LifeViewLinearLayout(Context context, AttributeSet attrs) {
            super(context, attrs);
            Log.d(TAG, "LifeViewGroup: 2");
        }

        public LifeViewLinearLayout(Context context, AttributeSet attrs, int defStyleAttr) {
            super(context, attrs, defStyleAttr);
            Log.d(TAG, "LifeViewGroup: 3");
        }

        @Nullable
        @Override
        protected Parcelable onSaveInstanceState() {
            Log.d(TAG, "onSaveInstanceState: view group");
            return super.onSaveInstanceState();
        }

        @Override
        protected void onRestoreInstanceState(Parcelable state) {
            super.onRestoreInstanceState(state);
            Log.d(TAG, "onRestoreInstanceState: view group");
        }
    }
}

File: demo/app/src/main/java/cn/dozyx/template/view/ViewLifeActivity.kt
package cn.dozyx.template.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Rect
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.View
import android.widget.LinearLayout
import androidx.appcompat.app.AppCompatActivity

import cn.dozyx.template.R
import timber.log.Timber

class ViewLifeActivity : LifeCycleActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_view_life)
    }

    class ViewLifeView : View {
        constructor(context: Context) : super(context) {
            Timber.d("ViewLifeView.constructor1")
        }

        constructor(context: Context,
                    attrs: AttributeSet?) : super(context, attrs) {
            Timber.d("ViewLifeView: constructor2")
        }

        constructor(context: Context,
                    attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
            Timber.d("ViewLifeView: constructor3")
        }

        override fun onSaveInstanceState(): Parcelable? {
            Timber.d("onSaveInstanceState: ")
            return super.onSaveInstanceState()
        }

        override fun onFinishInflate() {
            super.onFinishInflate()
            Timber.d("onFinishInflate: ")
        }

        override fun onAttachedToWindow() {
            super.onAttachedToWindow()
            Timber.d("onAttachedToWindow: ")
        }

        override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
            super.onLayout(changed, left, top, right, bottom)
            Timber.d("onLayout: ")
        }

        override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
            Timber.d("onMeasure: ")
            // 执行了两次
        }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            Timber.d("onDraw: ")
        }

        override fun onDetachedFromWindow() {
            super.onDetachedFromWindow()
            Timber.d("onDetachedFromWindow: ")
        }

        override fun onFocusChanged(gainFocus: Boolean, direction: Int,
                                    previouslyFocusedRect: Rect?) {
            super.onFocusChanged(gainFocus, direction, previouslyFocusedRect)
            Timber.d("onFocusChanged: $gainFocus")
        }

        override fun onRestoreInstanceState(state: Parcelable) {
            super.onRestoreInstanceState(state)
            Timber.d("onRestoreInstanceState: ")
        }

        override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
            super.onSizeChanged(w, h, oldw, oldh)
            Timber.d("ViewLifeView.onSizeChanged")
        }

        override fun dispatchDraw(canvas: Canvas?) {
            super.dispatchDraw(canvas)
            Timber.d("ViewLifeView.dispatchDraw")
        }

        override fun onTouchEvent(event: MotionEvent): Boolean {
            Timber.d("ViewLifeView.onTouchEvent ${MotionEvent.actionToString(event.action)}")
            return event.action == MotionEvent.ACTION_DOWN
//            return super.onTouchEvent(event)
        }
    }

}
class LifeViewLinearLayout : LinearLayout {
    constructor(context: Context) : super(context) {
        Timber.d("LifeViewGroup: 1")
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        Timber.d("LifeViewGroup: 2")
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        Timber.d("LifeViewGroup: 3")
    }

    override fun onSaveInstanceState(): Parcelable? {
        Timber.d("onSaveInstanceState: view group")
        return super.onSaveInstanceState()
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        super.onRestoreInstanceState(state)
        Timber.d("onRestoreInstanceState: view group")
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        Timber.d("LifeViewLinearLayout.onTouchEvent ${MotionEvent.actionToString(event.action)}")
        return super.onTouchEvent(event)
    }

    override fun onInterceptTouchEvent(ev: MotionEvent): Boolean {
        Timber.d("LifeViewLinearLayout.onInterceptTouchEvent ${MotionEvent.actionToString(ev.action)}")
        return super.onInterceptTouchEvent(ev)
//            return true
    }
}
